data_structure/segment_trees.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright 2020 @author Lasha Bukhnikashvili
+ * @file
+ * Program describes usage of a popular Data Structure: Segment Trees
+ * in Finding maximum number and its index on a given Interval
+ * in logarithmic = Log(N) Time using SegmentTree Data Structures
+ * with PreProcessing O(N * Log(N)) Time
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstdio>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+
+/**
+ * Each Node contains corresponding value
+ * in array and it's index
+ */
+struct Node {
+    int val;
+    int ind;
+} null;
+
+/**
+ * Data Structure stores initial array and its nodes,
+ * which has the 4th size of the array = 4*N
+ */
+struct SegmentTree {
+    int N;
+    std::vector<int> arr;
+    std::vector<Node> nodes;
+} ST;
+
+/**
+ * This is a recursive method, which
+ * Builds the Tree in ST struct variable.
+ * It should be called only once for each array.
+ * Works in O(N * log N)
+ */
+void BuildTree_Procedural(int v, int tl, int tr) {
+    if (tl == tr) {
+        ST.nodes[v].val = ST.arr[tl - 1], ST.nodes[v].ind = tl;
+    } else {
+        int tm = (tl + tr) / 2;
+        BuildTree_Procedural(2 * v, tl, tm);
+        BuildTree_Procedural(2 * v + 1, tm + 1, tr);
+        if (ST.nodes[2 * v].val >= ST.nodes[2 * v + 1].val)
+            ST.nodes[v] = ST.nodes[2 * v];
+        else
+            ST.nodes[v] = ST.nodes[2 * v + 1];
+    }
+}
+
+/**
+ * This method gets initial array and builds
+ * Tree Data Structure based on this array
+ * and stores it in ST struct variable
+ */
+void BuildTree(std::vector<int> arr) {
```

Comment:
per the Google C++ standards (that we are following for this repo) an optimal way to pass an un-modifiable class variable is as a const reference:
```suggestion
void BuildTree(const std::vector<int> &arr) {
```